### PR TITLE
feat: include sessionToken in onLiveQueryEvent

### LINKED ISSUE #1906: How to detect a user is offline when using liveQuery in a realtime chat app
I am using **live queries** to realize the realtime chat app, but I could not find **the method to detect a user is online or offline**. I am also confused with the `enter` and `leave` events.

```
subscription.on('enter', (object) => {
  console.log('object entered');
});
```

When an existing ParseObject's old value does not fulfill the ParseQuery but its new value fulfills the ParseQuery, you'll get this event. The object is the ParseObject which enters the ParseQuery. Its content is the latest value of the ParseObject.

```
subscription.on('leave', (object) => {
  console.log('object left');
});
```

When an existing ParseObject's old value fulfills the ParseQuery but its new value doesn't fulfill the ParseQuery, you'll get this event. The object is the ParseObject which leaves the ParseQuery. Its content is the latest value of the ParseObject.

I wonder when to trigger these two events. Thanks in advance!

### PR DESCRIPTION
**Pull request by dblythy**
Closes #1906.

If a sessionToken is passed onLiveQueryEvent, lookup the associated Parse.User, and set req.user.

### CODE REVIEW

File: src/triggers.js [hunk 1]
```diff
@@ -955,6 +972,9 @@ export async function maybeRunAfterEventTrigger(triggerType, className, request)
     request.original = Parse.Object.fromJSON(request.original);
   }
   request.user = await userForSessionToken(request.sessionToken);
+  if (!request.user) {
+    request.sessionToken = undefined;
+  }
```
Comment (dplewis) [inline src/triggers.js:977]: I believe these can be reverted.

File: src/triggers.js [hunk 2]
```diff
@@ -968,14 +988,10 @@ async function userForSessionToken(sessionToken) {
...
   if (!session) {
     return;
   }
-  const user = session.get('user');
-  if (!user) {
-    return;
-  }
-  await user.fetch({ useMasterKey: true });
-  return user;
+  return session.get('user');
```
Comment (dplewis) [inline src/triggers.js:996]: Nice cleanup!

File: src/triggers.js [hunk 3]
```diff
@@ -845,10 +845,21 @@ export function inflate(data, restObject) {
...
+    !_triggerStore ||
+    !_triggerStore[applicationId] ||
+    !_triggerStore[applicationId].LiveQuery ||
+    _triggerStore[applicationId].LiveQuery.length == 0
+  ) {
     return;
   }
+  if (data.sessionToken && !data.user) {
+    data.user = await userForSessionToken(data.sessionToken);
+    if (!data.user) {
```
Comment (dplewis) [inline src/triggers.js:859]: this can be removed

File: spec/ParseLiveQuery.spec.js [hunk 4]
```diff
@@ -6,6 +6,35 @@ const validatorFail = () => {
...
+        expect(sessionToken).toBe(requestedUser.getSessionToken());
+        done();
+      }
+    });
+    await requestedUser.signUp();
+    const query = new Parse.Query(TestObject);
+    await query.subscribe();
+    const client = await Parse.CoreManager.getLiveQueryController().getDefaultLiveQueryClient();
+    client.close();
+  });
```
Comment (dplewis) [inline spec/ParseLiveQuery.spec.js:37]: Can you add a space after the test `it()` for consistency

**Review by dplewis — COMMENTED** [review]
Just a comment and a nit

**Review by dplewis — APPROVED** [review]
Nice catch! I can't believe I missed this.

**Comment by dblythy** [comment]
Thanks for the review! Shoutout to @maxiqsoft for identifying the solution.

**Comment by dplewis** [comment]
@dblythy Actually I think passing the sessionToken should be enough. I don't know how I feel about doing a lookup for every live query event (could be hundreds of thousands of events). The user can easily get the user from the sessionToken.

**Comment by dblythy** [comment]
> @dblythy Actually I think passing the sessionToken should be enough. I don't know how I feel about doing a lookup for every live query event (could be hundreds of thousands of events)

This might be worth considering for `afterLiveQueryEvent` trigger too, which fires on  update, leave, enter, create delete, where `onLiveQueryEvent` fires on connect, subscribe, unsubscribe, ws_connect, ws_disconnect, ws_disconnect_error.

**Comment by mtrezza** [comment]
I think with the installation ID and the session token now both being available, there are now 2 ways of inferring the user. A user lookup should be done optionally if needed, due to the look-up costs.

**Comment by dplewis** [comment]
That’s probably why I didn’t add sessionToken here because installationId was enough.

**Comment by dblythy** [comment]
How do we feel about removing request.user from all LiveQuery triggers (beforeSubscribe, beforeConnect, afterLiveQueryEvent)?

We could potentially replace it with some sort of helper method such as:

const user = await Parse.Cloud.getUserFromSessionToken(sessionToken)

**Comment by mtrezza** [comment]
As I mentioned in the original tread #1906:

> it would currently be possible to infer the user from the installationId, but that is indirect and requires additional referencing. This solution on the other hand would allow to infer the user directly from the session object.

**Review by dplewis — APPROVED** [review]

**Pull request merged**